jgrapht-core/src/main/java/org/jgrapht/alg/cycle/AhujaOrlinSharmaCyclicExchangeLocalAugmentation.java

```diff
@@ -0,0 +1,343 @@
+     *
+     * @param graph the (improvement) graph on which to calculate the local augmentation
+     * @param lengthBound the upper bound for the length of cycles to detect
+     * @param labels the labels of the vertices encoding the subsets of vertices
+     */
+    public AhujaOrlinSharmaCyclicExchangeLocalAugmentation(Graph<V, E> graph, int lengthBound, Map<V, Integer> labels) {
+        this.graph = Objects.requireNonNull(graph, "Graph cannot be null");
+        this.lengthBound = lengthBound;
+        this.labels = Objects.requireNonNull(labels, "Labels cannot be null");
+    }
+
+    /**
+     * Calculates a valid subset-disjoint negative cycle.
+     * If there is no such cycle, it returns an empty LabeledPath instance with cost Double.MAX_VALUE.
+     *
+     * @return a valid subset-disjoint negative cycle encoded as LabeledPath
+     */
+    public GraphWalk<V, E> getLocalAugmentationCycle() {
+
+        int k = 1;
+        LabeledPath<V> cycleResult = new LabeledPath<>(lengthBound);
+
+        /*
+         * Store the path sets via a with Pair<V, V> as for the head and tail of the path such that the domination tests runs faster.
+         * Furthermore, use a TreeSet to sort for the cost of the paths such that the domination test stops optimally concerning the cost of the paths.
+         */
+        Map<Pair<V, V>, TreeSet<LabeledPath<V>>> pathsLengthK = new LinkedHashMap<>();
+        Map<Pair<V, V>, TreeSet<LabeledPath<V>>> pathsLengthKplus1 = new LinkedHashMap<>();
+
+        // initialize pathsLengthK for k = 1
+        for(E e : graph.edgeSet()) {
+            if(graph.getEdgeWeight(e) < 0) {
+                // initialize all paths of cost < 0
+                ArrayList<V> pathVertices = new ArrayList<>(lengthBound);
+                Set<Integer> pathLabels = new HashSet<>();
+                V sourceVertex = graph.getEdgeSource(e);
+                V targetVertex = graph.getEdgeTarget(e);
+                pathVertices.add(sourceVertex);
+                pathVertices.add(targetVertex);
+                pathLabels.add(labels.get(sourceVertex));
+                pathLabels.add(labels.get(targetVertex));
+                LabeledPath<V> path = new LabeledPath<>(pathVertices, graph.getEdgeWeight(e), pathLabels);
+
+                // add path to set of paths of length 1
+                addPathToPathStructure(pathsLengthK, path);
+            }
+        }
+
+        while (k < lengthBound && cycleResult.getCost() >= 0) {
+
+            // go through all valid paths of length k
+            for (TreeSet<LabeledPath<V>> tree : pathsLengthK.values()) {
+                for(LabeledPath<V> path : tree) {
+                    V head = path.getHead();
+                    V tail = path.getTail();
+
+                    // the path builds a valid negative cycle
+                    if (graph.containsEdge(tail, head)
+                            && path.getCost() + graph.getEdgeWeight(graph.getEdge(tail, head)) < 0
```

Comment:
Don't duplicate the computation of the new cost.  Also, I don't understand why both checks are needed.  We return as soon as we hit a negative cycle, so if the newly calculated cost is negative, then it is definitely less than the previously computed cycleResult cost.